### ISSUE TITLE
Failed: OpenSSL_add_all_digests while running backup job

### ISSUE THREAD

**Issue by deHoeninger** [duplicati/duplicati#2783]
- [x] I have searched open and closed issues for duplicates.

----------------------------------------

## Environment info

- **Duplicati version**: 2.0.2.9
- **Operating system**: Ubuntu 16.04
- **Backend**: ACD

## Description

After upgrading to version 2.9.2.9 I get the following error message once I'm starting a backup job.

```
Failed: OpenSSL_add_all_digests
Details: System.EntryPointNotFoundException: OpenSSL_add_all_digests
  at (wrapper managed-to-native) FasterHashing.InteropOpenSSL10:OpenSSL_add_all_digests ()
  at FasterHashing.OpenSSL10HashAlgorithmSHA256..ctor () <0x405d5950 + 0x0002b> in <filename unknown>:0 
  at FasterHashing.OpenSSL10HashAlgorithm.Create (System.String name) <0x405d56d0 + 0x000eb> in <filename unknown>:0 
  at FasterHashing.FasterHash.Create (System.String algorithm, Boolean allowfallback, HashImplementation implementation) <0x405d4900 + 0x00077> in <filename unknown>:0 
  at Duplicati.Library.Utility.HashAlgorithmHelper.Create (System.String name) <0x405d48d0 + 0x00013> in <filename unknown>:0 
  at Duplicati.Library.Main.Utility+Metahash..ctor (System.Collections.Generic.Dictionary`2 values, Duplicati.Library.Main.Options options) <0x405d4470 + 0x000f3> in <filename unknown>:0 
  at Duplicati.Library.Main.Utility.WrapMetadata (System.Collections.Generic.Dictionary`2 values, Duplicati.Library.Main.Options options) <0x405d4410 + 0x0003f> in <filename unknown>:0 
  at Duplicati.Library.Main.Operation.BackupHandler..ctor (System.String backendurl, Duplicati.Library.Main.Options options, Duplicati.Library.Main.BackupResults results) <0x405d4190 + 0x00087> in <filename unknown>:0 
  at Duplicati.Library.Main.Controller+<>c__DisplayClass16_0.<Backup>b__0 (Duplicati.Library.Main.BackupResults result) <0x405d2690 + 0x012a7> in <filename unknown>:0 
  at Duplicati.Library.Main.Controller.RunAction[T] (Duplicati.Library.Main.T result, System.String[]& paths, IFilter& filter, System.Action`1 method) <0x405b6000 + 0x00243> in <filename unknown>:0 
```

## Steps to reproduce

1. Run the backup-job

- **Actual result**:
Error-Message
- **Expected result**:
Backup Job succeeded without error message

**Comment by dgcom** [comment]
I think, you'll need to install OpenSSL yourself in order for Duplicati 2.0.2.9 to work - it relies on libssl.dll but does not include it in the package...

**Comment by deHoeninger** [comment]
Hi dgcom,

according to my package manager, I've openssl installed:

openssl is already the newest version (1.1.0f-2~ubuntu16.04.1+deb.sury.org+4).

**Comment by dgcom** [comment]
Take a look at the bottom of this readme: https://github.com/kenkendk/FasterHashing
It says that some Linux installations need additional step for mono.

**Comment by deHoeninger** [comment]
Thanks for the advice,
I've created the mentioned FasterHashing.dll.config file and it's seems to work now.

**Comment by deHoeninger** [comment]
I have to revert my positive feedback. Its still not working, on Ubuntu the lib seems to be located under 
/lib/x86_64-linux-gnu/libssl.so.1.0.0
I've created symlinks in /lib with names libssl.so.1.0.0 and libssl.so, but nothing worked.

Any further ideas?

**Comment by dgcom** [comment]
Are you getting the same exact exception? May be there is an issue with version mismatch...
But, no, I do not have more ideas... Maybe kenkendk can have more insight in this.

**Comment by deHoeninger** [comment]
Ok, after creating a symlink from /lib/x86_64-linux-gnu/libssl.so.1.0.0 to /usr/lib/libssl.so it seems to work now.

**Comment by enboig** [comment]
Same problem, creating a symlink worked

**Comment by valandil** [comment]
So is that something that can be fixed inside Duplicati, or will package maintainres for Linux distributions have to deal with this themselves?

**Comment by warwickmm** [comment]
I was having the same issue when running the unit tests on the `master` branch.  I have a symlink `/lib/libssl.so` that targets `/lib/libssl.so.1.1`.

[This page](https://www.openssl.org/docs/man1.1.0/crypto/OpenSSL_add_all_digests.html) seems to indicate that the `OpenSSL_add_all_digests` function was deprecated in OpenSSL 1.1.0.  Once I added an assembly config file that mapped to version 1.0.0, I no longer encountered any errors.

@deHoeninger, it sounds like you also got it working once you made your symlink target version 1.0.0 instead of 1.1.0?

**Comment by kenkendk** [comment]
If you have trouble with FasterHashing, it can be disabled with environment variables:
```
FH_DISABLE_OPENSSL11=1
FH_DISABLE_OPENSSL10=1
FH_DISABLE_APPLECC=1
FH_DISABLE_CNG=1
```

You can also select the version to use: 
`OpenSSL11`, `OpenSSL10`, `AppleCommonCrypto`, `CNG`, `Managed`, `openssl`

The `Managed` option is what Duplicati previously used, and that is bundled with Mono. The `openssl` is an alias for "Prefer OpenSSL 1.1, but use OpenSSL 1.0 if available"

To revert to the previous version, simply set:
```
FH_LIBRARY=managed
```

I will be investigating why it seems to fail with `OpenSSL 1.1` as reported in this issue.

**Comment by kenkendk** [comment]
Found and fixed the problem with OpenSSL 1.1, it will be included in the next canary build.

**Comment by joolsr** [comment]
Great news about the fix, I too succumbed to this problem when running Manjaro Linux, so its not just Ubuntu distros affected. Any Idea when you may be able to release the new build @kenkendk ?

**Comment by kenkendk** [comment]
I have included the new library in 78c5bdac0c79e70f7bd2b0b07562aa63ae481982

I will see if I can get a new build out tomorrow.

**Comment by kenkendk** [comment]
Follow up on this issue: in the latest version it will automatically detect the various names used by OpenSSL (`libssl.so`, `libssl.so.1.0` etc) on both Windows and Linux.
